aws/resource_aws_budgets_budget.go

```diff
@@ -0,0 +1,431 @@
+		BudgetName: &budgetName,
+		AccountId:  &accountID,
+	})
+	if err != nil {
+		return fmt.Errorf("delete budget failed: %v", err)
+	}
+
+	return nil
+}
+
+type flattenedBudgetsBudget struct {
+	name            *string
+	budgetType      *string
+	timeUnit        *string
+	costFilters     map[string][]*string
+	limitAmount     *string
+	limitUnit       *string
+	costTypes       map[string]bool
+	timePeriodStart *time.Time
+	timePeriodEnd   *time.Time
+}
+
+func expandBudgetsBudgetFlatten(budget *budgets.Budget) (*flattenedBudgetsBudget, error) {
+	if budget == nil {
+		return nil, fmt.Errorf("empty budget returned from budget output: %v", budget)
+	}
+
+	budgetLimit := budget.BudgetLimit
+	if budgetLimit == nil {
+		return nil, fmt.Errorf("empty limit in budget: %v", budget)
+	}
+
+	budgetCostTypes := budget.CostTypes
+	if budgetCostTypes == nil {
+		return nil, fmt.Errorf("empty CostTypes in budget: %v", budget)
+	}
+
+	costTypesMap := map[string]bool{
+		"include_credit":             *budgetCostTypes.IncludeCredit,
+		"include_other_subscription": *budgetCostTypes.IncludeOtherSubscription,
+		"include_recurring":          *budgetCostTypes.IncludeRecurring,
+		"include_refund":             *budgetCostTypes.IncludeRefund,
+		"include_subscription":       *budgetCostTypes.IncludeSubscription,
+		"include_support":            *budgetCostTypes.IncludeSupport,
+		"include_tax":                *budgetCostTypes.IncludeTax,
+		"include_upfront":            *budgetCostTypes.IncludeUpfront,
+		"use_blended":                *budgetCostTypes.UseBlended,
+	}
+	budgetTimePeriod := budget.TimePeriod
+	if budgetTimePeriod == nil {
+		return nil, fmt.Errorf("empty TimePeriod in budget: %v", budget)
+	}
+
+	budgetTimePeriodStart := budgetTimePeriod.Start
+	if budgetTimePeriodStart == nil {
+		return nil, fmt.Errorf("empty TimePeriodStart in budget: %v", budget)
+	}
+
+	budgetTimePeriodEnd := budgetTimePeriod.End
+	if budgetTimePeriodEnd == nil {
+		return nil, fmt.Errorf("empty TimePeriodEnd in budget: %v", budget)
+	}
+
+	return &flattenedBudgetsBudget{
+		name:            budget.BudgetName,
+		budgetType:      budget.BudgetType,
+		timeUnit:        budget.TimeUnit,
+		costFilters:     budget.CostFilters,
+		limitAmount:     budgetLimit.Amount,
+		limitUnit:       budgetLimit.Unit,
+		costTypes:       costTypesMap,
+		timePeriodStart: budgetTimePeriodStart,
+		timePeriodEnd:   budgetTimePeriodEnd,
+	}, nil
+}
+
+func convertCostFiltersToStringMap(costFilters map[string][]*string) map[string]string {
+	convertedCostFilters := make(map[string]string)
+	for k, v := range costFilters {
+		filterValues := make([]string, 0)
+		for _, singleFilterValue := range v {
+			filterValues = append(filterValues, *singleFilterValue)
+		}
+
+		convertedCostFilters[k] = strings.Join(filterValues, ",")
+	}
+
+	return convertedCostFilters
+}
+
+func expandBudgetsBudgetUnmarshal(d *schema.ResourceData) (*budgets.Budget, error) {
+	var budgetName string
```

Comment:
This budgetName logic should only live in the create function. It won't be necessary anywhere else when we call `d.Set("name", ...)` during read 👍